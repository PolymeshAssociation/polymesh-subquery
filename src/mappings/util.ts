import { decodeAddress } from '@polkadot/keyring';
import { Codec } from '@polkadot/types/types';
import { hexStripPrefix, u8aToHex, u8aToString } from '@polkadot/util';
import { SubstrateEvent, SubstrateExtrinsic } from '@subql/types';
import { Portfolio } from 'polymesh-subql/types/models/Portfolio';
import {
  AssetDocument,
  Compliance,
  Distribution,
  FoundType,
  SecurityIdentifier,
  TransferComplianceExemption,
  TransferManager,
  TransferRestrictionTypeEnum,
} from '../types';
import { Attributes } from './entities/common';

/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str[0].toLowerCase() + str.slice(1).replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`);

export const snakeToCamelCase = (value: string): string =>
  value
    .toLowerCase()
    .replace(/([-_][a-z])/g, group => group.toUpperCase().replace('-', '').replace('_', ''));

export const capitalizeFirstLetter = (str: string): string => str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s?.replace(/\0/g, '') || '';

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (text: string, exitOnFirst = false): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case '(':
      case '{':
      case '<': {
        nestedLevel++;
        break;
      }
      case ')':
      case '}':
      case '>': {
        nestedLevel--;
        break;
      }
      case ',': {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(`No top level comma found in ${text}, it probably isn't a map`);
  }
  return commas;
};

export const getOrDefault = <K, V>(map: Map<K, V>, key: K, getDefault: () => V): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};

export const serializeAccount = (item: Codec): string | null => {
  const s = item.toString();

  if (s.trim().length === 0) {
    return null;
  }
  return u8aToHex(decodeAddress(item.toString(), false, item.registry.chainSS58));
};

export const getFirstKeyFromJson = (item: Codec): string => {
  return Object.keys(item.toJSON())[0];
};

export const getFirstValueFromJson = (item: Codec): string => {
  return item.toJSON()[getFirstKeyFromJson(item)];
};

export const getTextValue = (item: Codec): string => {
  return item?.toString().trim().length === 0 ? null : item.toString().trim();
};

export const getNumberValue = (item: Codec): number => {
  return Number(getTextValue(item));
};

export const getDateValue = (item: Codec): Date => {
  return item.toString().trim().length === 0 ? null : new Date(Number(item.toString()));
};

export const getBigIntValue = (item: Codec): bigint => {
  return BigInt(getTextValue(item) || 0);
};

export const getBooleanValue = (item: Codec): boolean => {
  return JSON.parse(getTextValue(item));
};

export const hexToString = (input: string): string => {
  const hex = hexStripPrefix(input);
  let str = '';
  for (let i = 0; i < hex.length; i += 2)
    str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
  return removeNullChars(str);
};

export const getSigner = (extrinsic: SubstrateExtrinsic): string => {
  return hexStripPrefix(
    u8aToHex(
      decodeAddress(
        extrinsic.extrinsic.signer.toString(),
        false,
        extrinsic.extrinsic.registry.chainSS58
      )
    )
  );
};

/**
 * Parses a raw Asset Document
 */
export const getDocValue = (
  doc: Codec
): Pick<AssetDocument, 'name' | 'link' | 'contentHash' | 'type' | 'filedAt'> => {
  const {
    uri: link,
    content_hash: documentHash,
    name,
    doc_type: type,
    filing_date: filedAt,
  } = JSON.parse(doc.toString());

  const hashType = Object.keys(documentHash)[0];

  return {
    name,
    link,
    contentHash: documentHash[hashType],
    type,
    filedAt,
  };
};

export const getSecurityIdentifiers = (item: Codec): SecurityIdentifier[] => {
  const identifiers = JSON.parse(item.toString());
  return identifiers.map(i => {
    const type = Object.keys(i)[0];
    return {
      type,
      value: i[type],
    };
  });
};

/**
 * Parses a Vec<AssetCompliance>
 */
export const getComplianceRulesValue = (
  requirements: Codec
): Pick<Compliance, 'complianceId' | 'data'>[] => {
  return JSON.parse(JSON.stringify(requirements));
};

/**
 * Parses AssetCompliance
 */
export const getComplianceValue = (
  compliance: Codec
): Pick<Compliance, 'complianceId' | 'data'> => {
  const { id, ...data } = JSON.parse(compliance.toString());
  return {
    complianceId: Number(id),
    data,
  };
};

/**
 * Parses AssetTransferManager
 */
export const getTransferManagerValue = (
  manager: Codec
): Pick<TransferManager, 'type' | 'value'> => {
  const { countTransferManager, percentageTransferManager } = JSON.parse(JSON.stringify(manager));

  if (countTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Count,
      value: Number(countTransferManager),
    };
  }

  if (percentageTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Percentage,
      value: Number(percentageTransferManager),
    };
  }

  throw new Error('Unknown transfer restriction type found');
};

export const getExemptKeyValue = (
  item: Codec
): Omit<Attributes<TransferComplianceExemption>, 'exemptedEntityId'> => {
  const {
    asset: { ticker },
    op: opType,
    claimType,
  } = JSON.parse(item.toString());
  return {
    assetId: hexToString(ticker),
    opType,
    claimType,
  };
};

export const getExemptionsValue = (exemptions: Codec): string[] => {
  return exemptions.toJSON() as string[];
};

export const logFoundType = (type: string, rawType: string): void => {
  FoundType.create({ id: type, rawType }).save();
};

export const END_OF_TIME = BigInt('253402194600000');

export function addIfNotIncludes<T>(arr: T[], item: T): void {
  if (!arr.includes(item)) {
    arr.push(item);
  }
}

interface MeshPortfolio {
  did: string;
  kind: {
    user: number;
  };
}

const meshPortfolioToPortfolio = ({
  did,
  kind: { user: number },
}: MeshPortfolio): Pick<Portfolio, 'identityId' | 'number'> => ({
  identityId: did,
  number: number || 0,
});

export const getPortfolioValue = (item: Codec): Pick<Portfolio, 'identityId' | 'number'> => {
  const meshPortfolio = JSON.parse(item.toString());
  return meshPortfolioToPortfolio(meshPortfolio);
};

export const getCaIdValue = (item: Codec): Pick<Distribution, 'localId' | 'assetId'> => {
  const { local_id: localId, ticker } = JSON.parse(item.toString());
  return { localId, assetId: hexToString(ticker) };
};

export interface LegDetails {
  from: Pick<Portfolio, 'identityId' | 'number'>;
  to: Pick<Portfolio, 'identityId' | 'number'>;
  ticker: string;
  amount: bigint;
}

export const getLegsValue = (item: Codec): LegDetails[] => {
  const legs = JSON.parse(item.toString());
  return legs.map(({ from: fromPortfolio, to: toPortfolio, asset: ticker, amount }) => ({
    from: meshPortfolioToPortfolio(fromPortfolio),
    to: meshPortfolioToPortfolio(toPortfolio),
    ticker: hexToString(ticker),
    amount: getBigIntValue(amount),
  }));
};

export const getSignerAddress = (event: SubstrateEvent): string => {
  let signer: string;
  if (event.extrinsic) {
    signer = getSigner(event.extrinsic);
  }
  return signer;
};

export const getDistributionValue = (
  item: Codec
): Pick<
  Distribution,
  'portfolioId' | 'currency' | 'perShare' | 'amount' | 'remaining' | 'paymentAt' | 'expiresAt'
> => {
  const { from, currency, per_share, amount, remaining, payment_at, expires_at } = JSON.parse(
    item.toString()
  );
  const { identityId, number } = meshPortfolioToPortfolio(from);
  return {
    portfolioId: `${identityId}/${number}`,
    currency: hexToString(currency),
    perShare: getBigIntValue(per_share),
    amount: getBigIntValue(amount),
    remaining: getBigIntValue(remaining),
    paymentAt: getBigIntValue(payment_at),
    expiresAt: getBigIntValue(expires_at || END_OF_TIME),
  };
};

export const logError = (message: string): void => {
  logger.error(message);
};
