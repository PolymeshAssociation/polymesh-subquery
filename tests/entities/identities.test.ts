import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('accounts', () => {
  it('should return number of accounts associated with a given identity', async () => {
    const identityId = '0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e';
    const res = await query({
      query: gql`
        query {
          accounts(filter: {
              identityId: {
                equalTo: "${identityId}"
              }
            }) {
            totalCount
          }
        }
      `,
    });

    return expect(res?.data?.accounts?.totalCount).toEqual(2);
  });

  it('should return just total count', async () => {
    const res = await query({
      query: gql`
        query {
          accounts {
            totalCount
          }
        }
      `,
    });

    expect(res?.data).toMatchSnapshot();
  });

  it('should return first 10 accounts as there are no filters', async () => {
    const result = await query({
      query: gql`
        query {
          accounts(first: 10) {
            totalCount
            nodes {
              address
              identityId
              eventId
              permissionsId
              createdBlockId
              datetime
            }
          }
        }
      `,
    });

    expect(result?.errors).toBeFalsy();

    expect(result?.data).toMatchSnapshot();
  });

  it('should return filtered accounts along with their permissions', async () => {
    const identityId = '0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e';
    const result = await query({
      query: gql`
        query {
          accounts(
            filter: {
              identityId: {
                equalTo: "${identityId}"
              }
            }
          ) {
            nodes {
              id
              address
              identityId
              eventId
              permissionsId
              createdBlockId
              datetime
              identity {
                did
                primaryAccount
                secondaryKeysFrozen
              }
              permissions {
                assets
                portfolios
                transactions
                transactionGroups
              }
            }
          }
        }
      `,
    });

    expect(result?.errors).toBeFalsy();

    expect(result?.data).toMatchSnapshot();
  });
});

describe('permissions', () => {
  it('should return permissions for a given account address', async () => {
    const address = '5EYxLuFdbD99jn2BsZ3f1rMoa3raDB8TAnTLX3VMXhQoJyrv';
    const res = await query({
      query: gql`
      query {
        permissions(filter: { id: { equalTo: "${address}"}}) {
          nodes {
            nodeId
            id
            assets
            portfolios
            transactions
            transactionGroups
            createdBlockId
            updatedBlockId
            datetime
          }
        }
      }
      `,
    });

    expect(res?.errors).toBeFalsy();

    expect(res?.data).toMatchSnapshot();
  });
});

describe('identities', () => {
  it('should return identity details along with accounts and their permissions', async () => {
    const identityId = '0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e';
    const res = await query({
      query: gql`
      query {
        identities(
          filter: {
            did: {
              equalTo: "${identityId}"
            }
          }
        ) {
          nodes {
            did
            primaryAccount
            secondaryKeysFrozen
            eventId
            createdBlockId
            updatedBlockId
            datetime
            createdBlockId
            updatedBlockId
            secondaryAccounts(orderBy: [ADDRESS_ASC]) {
              nodes {
                address
                eventId
                createdBlockId
                permissions {
                  assets
                  portfolios
                  transactions
                  transactionGroups
                  createdBlockId
                  updatedBlockId
                  datetime
                }
              }
            }
          }
        }
      }
      `,
    });

    expect(res?.errors).toBeFalsy();

    expect(res?.data).toMatchSnapshot();
  });

  it('should return a list of identities whose secondary keys are frozen', async () => {
    const result = await query({
      query: gql`
        query {
          identities(filter: { secondaryKeysFrozen: { equalTo: true } }) {
            nodes {
              did
            }
          }
        }
      `,
    });

    expect(result?.errors).toBeFalsy();

    expect(result?.data?.identities?.nodes?.length).toEqual(0);
  });
});
