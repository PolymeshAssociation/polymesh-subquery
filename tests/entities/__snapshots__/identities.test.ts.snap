// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts should return filtered accounts along with their permissions 1`] = `
Object {
  "accounts": Object {
    "__typename": "AccountsConnection",
    "nodes": Array [
      Object {
        "__typename": "Account",
        "address": "5EYxLuFdbD99jn2BsZ3f1rMoa3raDB8TAnTLX3VMXhQoJyrv",
        "createdBlockId": "345",
        "datetime": "2021-10-06T12:34:57.5",
        "eventId": "SecondaryKeysAdded",
        "id": "5EYxLuFdbD99jn2BsZ3f1rMoa3raDB8TAnTLX3VMXhQoJyrv",
        "identity": Object {
          "__typename": "Identity",
          "did": "0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e",
          "primaryAccount": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
          "secondaryKeysFrozen": false,
        },
        "identityId": "0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e",
        "permissions": Object {
          "__typename": "Permission",
          "assets": Object {
            "type": "these",
            "values": Array [],
          },
          "portfolios": Object {
            "type": "these",
            "values": Array [],
          },
          "transactionGroups": Array [],
          "transactions": Object {
            "type": "these",
            "values": Array [],
          },
        },
        "permissionsId": "5EYxLuFdbD99jn2BsZ3f1rMoa3raDB8TAnTLX3VMXhQoJyrv",
      },
      Object {
        "__typename": "Account",
        "address": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
        "createdBlockId": "341",
        "datetime": "2021-10-06T12:34:55.5",
        "eventId": "DidCreated",
        "id": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
        "identity": Object {
          "__typename": "Identity",
          "did": "0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e",
          "primaryAccount": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
          "secondaryKeysFrozen": false,
        },
        "identityId": "0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e",
        "permissions": Object {
          "__typename": "Permission",
          "assets": null,
          "portfolios": null,
          "transactionGroups": Array [],
          "transactions": null,
        },
        "permissionsId": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
      },
    ],
  },
}
`;

exports[`accounts should return first 10 accounts as there are no filters 1`] = `
Object {
  "accounts": Object {
    "__typename": "AccountsConnection",
    "nodes": Array [
      Object {
        "__typename": "Account",
        "address": "5C7abMfbhWZGrzfnTeS7sTnY9bm7Ec8c1FBAkRdjw8iT41ZD",
        "createdBlockId": "186",
        "datetime": "2021-10-06T12:33:38",
        "eventId": "DidCreated",
        "identityId": "0x0096ea2b9174cbe71c1d3125850b06b2a6e6007cdb3cb33fc371f07154847135",
        "permissionsId": "5C7abMfbhWZGrzfnTeS7sTnY9bm7Ec8c1FBAkRdjw8iT41ZD",
      },
      Object {
        "__typename": "Account",
        "address": "5CDKC4cEo9gTRnBNZo4fpDS6ov1pcmngmf6QhNf9HzfBKPk3",
        "createdBlockId": "149",
        "datetime": "2021-10-06T12:33:19.5",
        "eventId": "DidCreated",
        "identityId": "0x70da84f285540a6174594f6fd69c7facf092cd29210f1b93ee3f4915c4c8f86c",
        "permissionsId": "5CDKC4cEo9gTRnBNZo4fpDS6ov1pcmngmf6QhNf9HzfBKPk3",
      },
      Object {
        "__typename": "Account",
        "address": "5CGTLF38iqat5h8UmDYNyTDu4PD6bUuqJ5LjWJQYwzQayeX3",
        "createdBlockId": "150",
        "datetime": "2021-10-06T12:33:20",
        "eventId": "DidCreated",
        "identityId": "0x9e0f63e20537f9e5dddd47540266080e25177a329a1b5a0bff9c950fe7a7da1d",
        "permissionsId": "5CGTLF38iqat5h8UmDYNyTDu4PD6bUuqJ5LjWJQYwzQayeX3",
      },
      Object {
        "__typename": "Account",
        "address": "5CM8cBoq3rqSDm8SUMmVsf9gZUtHYb984JAf4eKYmUtAiMhP",
        "createdBlockId": "125",
        "datetime": "2021-10-06T12:33:07.5",
        "eventId": "DidCreated",
        "identityId": "0xc5e2d554233da63d509ee482dbeed0ddec94dc1d0b45ebfdcdc48bd0928222b1",
        "permissionsId": "5CM8cBoq3rqSDm8SUMmVsf9gZUtHYb984JAf4eKYmUtAiMhP",
      },
      Object {
        "__typename": "Account",
        "address": "5CRvGY4utGxbmrS5uoxUkrfgyC38L48pixh6dVhpwLznsLZe",
        "createdBlockId": "176",
        "datetime": "2021-10-06T12:33:33",
        "eventId": "DidCreated",
        "identityId": "0x901facd0a58355c6cd5bd5cfe0179a8b49a721ea8bc2b2c06f43c17938edec3f",
        "permissionsId": "5CRvGY4utGxbmrS5uoxUkrfgyC38L48pixh6dVhpwLznsLZe",
      },
      Object {
        "__typename": "Account",
        "address": "5CXYpJxGyMKcJX7qjVXjJcA8w9UbsonRcLn32h6riHtMEB1b",
        "createdBlockId": "229",
        "datetime": "2021-10-06T12:33:59.5",
        "eventId": "DidCreated",
        "identityId": "0x9b8597fea4fa31fb8a1367a0230e3e52391a401aa0a5c79c712e7a9c24643ef3",
        "permissionsId": "5CXYpJxGyMKcJX7qjVXjJcA8w9UbsonRcLn32h6riHtMEB1b",
      },
      Object {
        "__typename": "Account",
        "address": "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
        "createdBlockId": "365",
        "datetime": "2021-10-06T12:35:07.5",
        "eventId": "DidCreated",
        "identityId": "0xad05ae379784a960d841080123d88b7431bf957a596f0afcea1851902a5f7944",
        "permissionsId": "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
      },
      Object {
        "__typename": "Account",
        "address": "5CkwnyhRgn6BuEiYx6jurDDvT6c4MtVnEtDfQbTWAUxbL7Sm",
        "createdBlockId": "158",
        "datetime": "2021-10-06T12:33:24",
        "eventId": "DidCreated",
        "identityId": "0x56a91c10f2368b30670b7bea260928f0291387abfb75b5953cd722917423bf01",
        "permissionsId": "5CkwnyhRgn6BuEiYx6jurDDvT6c4MtVnEtDfQbTWAUxbL7Sm",
      },
      Object {
        "__typename": "Account",
        "address": "5CrMp8ekpRUDBu1Ytyuqs22B6zXMpH7L2Drh5kALckYD34Ca",
        "createdBlockId": "67",
        "datetime": "2021-10-06T12:32:38.5",
        "eventId": "DidCreated",
        "identityId": "0x4dfc4fc610db8b59e29bb3df5b0a0b097fd5df5934dcdba475c470e302f205ea",
        "permissionsId": "5CrMp8ekpRUDBu1Ytyuqs22B6zXMpH7L2Drh5kALckYD34Ca",
      },
      Object {
        "__typename": "Account",
        "address": "5CrPPidjxv1aaupnq8atgsVpXNUmnZaqfyzhPcBTmkXqqsSt",
        "createdBlockId": "111",
        "datetime": "2021-10-06T12:33:00.5",
        "eventId": "DidCreated",
        "identityId": "0x6f6e1b8b1554b91f186a5040bad955e00b526e7d27dab20c0f4f4fdedd94cd0b",
        "permissionsId": "5CrPPidjxv1aaupnq8atgsVpXNUmnZaqfyzhPcBTmkXqqsSt",
      },
    ],
    "totalCount": 41,
  },
}
`;

exports[`accounts should return just total count 1`] = `
Object {
  "accounts": Object {
    "__typename": "AccountsConnection",
    "totalCount": 41,
  },
}
`;

exports[`identities should return identity details along with accounts and their permissions 1`] = `
Object {
  "identities": Object {
    "__typename": "IdentitiesConnection",
    "nodes": Array [
      Object {
        "__typename": "Identity",
        "createdBlockId": "341",
        "datetime": "2021-10-06T12:34:55.5",
        "did": "0xc5ccde150ee81ea66cb9a7d49250ba6590087504322412bef75093b68134c59e",
        "eventId": "DidCreated",
        "primaryAccount": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
        "secondaryAccounts": Object {
          "__typename": "AccountsConnection",
          "nodes": Array [
            Object {
              "__typename": "Account",
              "address": "5EYxLuFdbD99jn2BsZ3f1rMoa3raDB8TAnTLX3VMXhQoJyrv",
              "createdBlockId": "345",
              "eventId": "SecondaryKeysAdded",
              "permissions": Object {
                "__typename": "Permission",
                "assets": Object {
                  "type": "these",
                  "values": Array [],
                },
                "createdBlockId": "345",
                "datetime": "2021-10-06T12:34:57.5",
                "portfolios": Object {
                  "type": "these",
                  "values": Array [],
                },
                "transactionGroups": Array [],
                "transactions": Object {
                  "type": "these",
                  "values": Array [],
                },
                "updatedBlockId": "345",
              },
            },
            Object {
              "__typename": "Account",
              "address": "5F77TQS5VkQP7UuEuR9YjqFgyuc9tWgQjzjTi3WMKsbgTrfG",
              "createdBlockId": "341",
              "eventId": "DidCreated",
              "permissions": Object {
                "__typename": "Permission",
                "assets": null,
                "createdBlockId": "341",
                "datetime": "2021-10-06T12:34:55.5",
                "portfolios": null,
                "transactionGroups": Array [],
                "transactions": null,
                "updatedBlockId": "341",
              },
            },
          ],
        },
        "secondaryKeysFrozen": false,
        "updatedBlockId": "341",
      },
    ],
  },
}
`;

exports[`permissions should return permissions for a given account address 1`] = `
Object {
  "permissions": Object {
    "__typename": "PermissionsConnection",
    "nodes": Array [
      Object {
        "__typename": "Permission",
        "assets": Object {
          "type": "these",
          "values": Array [],
        },
        "createdBlockId": "345",
        "datetime": "2021-10-06T12:34:57.5",
        "id": "5EYxLuFdbD99jn2BsZ3f1rMoa3raDB8TAnTLX3VMXhQoJyrv",
        "nodeId": "WyJwZXJtaXNzaW9ucyIsIjVFWXhMdUZkYkQ5OWpuMkJzWjNmMXJNb2EzcmFEQjhUQW5UTFgzVk1YaFFvSnlydiJd",
        "portfolios": Object {
          "type": "these",
          "values": Array [],
        },
        "transactionGroups": Array [],
        "transactions": Object {
          "type": "these",
          "values": Array [],
        },
        "updatedBlockId": "345",
      },
    ],
  },
}
`;
